### PR TITLE
Adding IterativeLinearSolver and Tutorial

### PR DESCRIPTION
**Pull request by mahegz**
Adding IterativeLinearSolver which builds on was_sol to allow a user to implement algorithms relying on iteratively solving a series of linear problems. The main motivation is to solve linear problems with sparsity or class-based regularization, but the implemented solver allows more flexibility.    

Also adding a tutorial that, in the context of domain adaptation, uses IterativeLinearSolver to solve regularized linear problems using generalized conditional gradient.

I did not implement tests but I can add them if the design approach and the code are alright.

### CODE REVIEW

**Comment by michalk8** [comment]
Hi @mahegz , thanks a lot for the PR! Could you please add the `domain_adaptation_with_class_regularization.ipynb` somewhere in an appropriate place in https://github.com/ott-jax/ott/blob/main/docs/tutorials/index.rst? Also, I think the math is not be rendering, that's why the docs linter fails (apart from some minor issues I can fix later) - can you please fix the math in the notebook?
After this is done, will give this PR a proper review!

File: src/ott/solvers/iterative_linear.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+        linear_state=linear_sol,
+        prob_state=prob_state,
+        costs=costs,
+        linear_convergence=linear_convergence,
+        errors=errors,
+    )
+
+
+@jax.tree_util.register_pytree_node_class
+class IterativeLinearSolver(was_solver.WassersteinSolver):
```
Comment (michalk8) [inline src/ott/solvers/iterative_linear.py:101]: @mahegz sorry for the delay

After taking a closer look at the PR and the notebook, I think many of the things in the notebook can be moved from there into the code (esp. the `gcg`). Here's what I propose:
- rather than aiming for a very general solution, I think it will be better to have a good GCG implementation in OTT-JAX + the regularizations you have implemented in the notebook (Laplace + group lasso)
- the user should be able to easily pass their own regularization function when initializing the solver as a callable
- no need to have such general/complex `__call__` method, it should just accept the `LinearProblem` + optional initialization

This should give us a good start to have a GCG implementation, which we can then showcase in the notebook using the DA as an example - let me know what you think!

File: src/ott/solvers/iterative_linear.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+          "If initial solver state is None, init_linear_pb and \
+            init_prob_state should be provided"
+      )
+    if init_state is None:
+      init_state = self.init_state(init_linear_pb, init_prob_state, rng)
+
+    return iterations(self, init_state, next_linear_pb, new_pstate_cost)
+
+  @property
+  def warm_start(self) -> bool:
```
Comment (michalk8) [inline src/ott/solvers/iterative_linear.py:224]: I don't think this is needed, can be removed (in the original paper, I haven't seen warm starts).

**Review by michalk8 — CHANGES_REQUESTED** [review]
Thanks a lot @mahegz and apologies for the late review again!

I really like this contribution, would just focus on GCG first with DA as an application (means moving stuff from the notebook).

File: src/ott/solvers/iterative_linear.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+  inner iterations, so the user can display the error at each
+  iteration, e.g., using a progress bar.
+  See :func:`~ott.utils.default_progress_fn` for a basic implementation.
+  kwargs_init: auxilary key arguments
+  """
+
+  def __init__(
+      self,
+      *args: Any,
+      warm_start: Optional[bool] = None,
```
Comment (michalk8) [inline src/ott/solvers/iterative_linear.py:120]: I don't think any of these arguments should be exposed here (as it currently stands):
- warm_start can be removed
- `progress_fn` - usually is passed to the inner solver in `**kwargs`
- `kwargs_init` - would pass in `__call__` as `**kwargs`

File: src/ott/solvers/iterative_linear.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+        errors=errors,
+    )
+
+
+@jax.tree_util.register_pytree_node_class
+class IterativeLinearSolver(was_solver.WassersteinSolver):
+  """Implements probems requiring iterative calls to (low_rank) sinkhorn solver.
+
+  Args:
+  args: Positional arguments for
```
Comment (michalk8) [inline src/ott/solvers/iterative_linear.py:105]: The docs needs to be correctly indented as:
```rST
Args:
  args: Positional...
```

File: src/ott/solvers/iterative_linear.py (hunk 1 above)
Comment (michalk8) [inline src/ott/solvers/iterative_linear.py:101]: Could you also please add this solver to `docs/solvers/index.rst`?

Comment (mahegz) [reply src/ott/solvers/iterative_linear.py:101]: Thanks for the review! I agree that having an easy to use GCG solver is a good idea. 

For now, I hope to keep the more general `IterativeLinearSolver` I wrote and use it to implement an easy to use GCG.  The alternative would be to transform that solver into just a GCG solver. Let me know which direction you prefer.

Comment (michalk8) [reply src/ott/solvers/iterative_linear.py:101]: > For now, I hope to keep the more general IterativeLinearSolver I wrote and use it to implement an easy to use GCG. The alternative would be to transform that solver into just a GCG solver. Let me know which direction you prefer.

I'd prefer have it transformed into a GCG solver, we can think about the next step (more abstraction) later!

**Comment by marcocuturi** [comment]
Hello! Any news on this PR?

**Comment by mahegz** [comment]
> Hello! Any news on this PR?

Hi! Sorry for the delay I was quite busy in the last two weeks. 
I took note of @michalk8 review and plan to integrate his comments sometime in the next two weeks.

**Comment by mahegz** [comment]
> View / edit / reply to [this conversation](https://app.reviewnb.com/ott-jax/ott/pull/420/discussion/#comment-1766359887) on ReviewNB
> 
> _michalk8 commented on 2023-10-17T12:53:06Z_ ----------------------------------------------------------------
> 
> As discussed in the other comments, this should be part of the GCG solver and user should not implement this.

I moved GCG into the code. For the regularization functions, I am not entirely sure how you prefer to have it implemented. The regularization functions can be based on different types of auxiliary data such as the source/target labels.  We can have the two regularisation functions as standalone components in the GCG file. The alternative is to have the user pass a string argument, but that would require the user to supply different inputs to GCG based on the chosen regularizer.

**Comment by michalk8** [comment]
I will be closing this PR, as it's inactive and currently, I think we'd need a better abstraction for solvers to implement GCG.
